---
title: "Decovlution of differentiating adipoyctes"
author: "Mahmoud Shaaban"
date: "2/3/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE, error = FALSE, cache = TRUE)
```

# Overview

The code in this document reproduces the figures and tables for the manuscript
titled "Decovlution of differentiating adipoyctes." We start by obtaining the
the input data, loading the required libraries and explaining the structure of
the analysis workflow. The rest of the document shows the code to generate the
output of the analysis including the intermediary data objects and the final
figures and tables.

# Obtaining the data

This analysis is based on gene expression data from a time course experiment
of 3T3-L1 induced for differentiation using MDI and sampled at multiple time
points. Another similar dataset of induced human primary adipocytes using the
same protocol. Finally, a dataset of gene expression of induced 3T3-L1 with
chemical course perturbation is used.

The following code chunck downloads the datasets to the `data/` directory. The
contents of this directory include:

- `adipo_counts.rds`: A `SummarizedExperiment` containing the RNA-Seq gene 
expression of the time course experiment of 3T3-L1 pre-adipocytes

- `primary_adipocyte.rds`: An `ExpressionSet` containing the microarrays gene

expression of the time course experiment of human primary adipocytes

- `perturbations_md.csv`: A text file containing the metadata of the
perturbation experiments

- `pharmacological_perturbation.rds`: A `SummarizedExperiment` object 
containing the microarrays gene expression with induction and perturbations of
3T3-L1 pre-adipocytes

```{r get_data}
# download data files
if (!file.exists('data.zip')) {
    download.file(
    'https://ndownloader.figshare.com/articles/12840101/versions/1',
    destfile = 'data.zip'
)
    unzip('data.zip', exdir = 'data')
}
```

# Loading libraries

The following code chunk load the required R packages. These packages fall in
one of four categories: data management, analysis, visualization and annotation.

```{r load_libraries}
# data management
library(SummarizedExperiment)
library(reshape2)
library(tidyverse)

# analysis packages
library(clusterProfiler)
library(debCAM)
library(pdist)
library(limma)
library(PSEA)
library(deSolve)
library(scales)
library(DESeq2)
library(usedist)
library(fgsea)
library(Hmisc)
library(pcr)

# data visualization
library(xtable)
library(ggalt)
library(gridExtra)
library(png)
library(ComplexHeatmap)
library(cowplot)
library(circlize)
library(GGally)

# annotation packages
library(org.Mm.eg.db)
library(GO.db)
```

# Prepaing the gene expression data
## Differentiating 3T3-L1 pre-adipocytes

The dataset `adipo_counts` consists of RNA-Seq read counts in mm10 known genes.
The samples were generated from MDI-induced 3T3-L1 adipocytes and profiled at 
different times points.

```{r mouse_cell}
# load gene counts
adipo_counts <- read_rds('data/adipo_counts.rds')

# extract the expression matrix
mat <- assay(adipo_counts)
dim(mat)

# extract time info
time <- adipo_counts$time
table(time)
```

Now we need to subset the data to the differentiation stages with sufficient
samples. Then we transform the `SummarizedExperiment` object into `DESeq` object
to apply the transformation `vst`.

```{r make_deseq_object}
# filter low samples stage
se <- adipo_counts[, adipo_counts$stage %in% c(0, 1, 3)]

# transform the data
dds <- DESeqDataSet(se, ~stage)
dds_transform <- vst(dds)
```

## Differentiating human primary adipocytes

The dataset `primary_se` was generated by treating human sub-cutanous fat cells
using MDI. The samples were collected before and 7 days after treatment and
profiled using microarrays.

```{r human_cell}
# load primary adipocytes
primary_se <- read_rds('data/primary_adipocyte.rds')
dim(primary_se)

# inverse the log2 transformation
primary_mat <- 2 ^ na.omit(exprs(primary_se))

# extract group information
group <- primary_se$group
table(group)
```

## Drug perturbation of differentiating 3T3-L1

The dataset `pharma_mats` is the expression matrices composed of several
experiments. In each experiment, MDI-induced 3T3-L1 cells were treated with a
drug/compound and compared to non-treated controls.

```{r perturbations}
# load selected sample info
selected_samples <- read_csv('data/perturbations_md.csv')
selected_list <- split(selected_samples, selected_samples$series_id)

table(selected_samples$treatment_type)

# load and subset data
pharma <- read_rds('data/pharmacological_perturbation.rds')
pharma_mats <- map(selected_list, function(x) {
    se <- pharma[, colnames(pharma) %in% x$sample_id]
    na.omit(assay(se))
})

map(pharma_mats, dim)

# reverse the log2 transformation
pharma_mats$GSE42220 <- 2^pharma_mats$GSE42220
pharma_mats$GSE64075 <- 2^pharma_mats$GSE64075
```

# Data analysis 
## Multi-dimensional scaling (MDS) analysis

In this section, we apply multidimensional scaling (MDS) analysis to 
differentiating adipocytes. The distances between the samples are calculated
using `dist` and passed to `cmdscale`

```{r perform_mds}
# dissimilarity of samples
d <- dist(t(assay(dds_transform)))

# multidimensional scaling
mds <- cmdscale(d)

# format mds table
mds_df <- mutate(as.data.frame(mds),
              stage = ifelse(se$stage == 3, 'S2', paste0('S', se$stage)))
```

This figure shows the first two dimensions of the MDS colored by stage of 
differentiation. In other words, the samples are laid out on two dimensions
plane in a way to approximate the distances among them.

```{r plot_mds}
# make mds plot
mds_df %>%
    ggplot(aes(x = V1, y = V2, color = stage)) +
    geom_point() +
    geom_encircle() +
    labs(x = 'Dimension One',
         y = 'Dimension Two',
         color = 'Differentiation') +
    theme_bw() +
    theme(legend.position = 'none',
          panel.grid = element_blank(),
          panel.border = element_rect(size = 1.2)) +
    annotate('text', -50, 50, label = 'S0', color = 'red') +
    annotate('text', 90, 50, label = 'S1', color = 'blue') +
    annotate('text', 50, -50, label = 'S2', color = 'darkgreen')
```

A simple diagnostic of this analysis is to calculate the distances between
samples within and inbetween each groups. The groups can be compared using
`t.test`.

```{r more_distances}
# calculate full matrix distances
dist_groups <- dist_groups((d), dds_transform$stage)

# show the distribution of the distances
# hist(d)
hist(dist_groups$Distance)

# split the distances by labels
dist_group_list <- split(dist_groups$Distance, dist_groups$Label)

# test 0 and 1 samples are farther than samples withing 0
t.test(dist_group_list$`Between 0 and 1`,
       dist_group_list$`Within 0`,
       alternative = 'greater')

# test 0 and 3 samples are farther than samples withing 0
t.test(dist_group_list$`Between 0 and 3`,
       dist_group_list$`Within 0`,
       alternative = 'greater')
```

Other diagnostics for the MDS is to show the godness-of-fit which shows how
much of the variance is explained by the two dimensions. 

```{r more_mds}
# call mds with diagnostics
mds2 <- cmdscale(dist(t(assay(dds_transform))),
                 k = 2,
                 list. = TRUE,
                 eig = TRUE)

# eigenvalues
barplot(mds2$eig)

# gof
mds2$GOF

# dissimilarity
dissim <- cmdscale(1/(cor((assay(dds_transform)))),
                   k = 2,
                   eig = TRUE)

# stress
k <- 2:10
stress <- map(k, function(x) MASS::isoMDS(d, k = x)$stress)
plot(k, unlist(stress))
```

## Deconvolution analysis

```{r deconvolve_mouse}
# set a seed
set.seed(111)

# run deconvolution
rCAM <- CAM(mat,
            K = 2:5,
            thres.low = 0.30,
            thres.high = 0.95,
            cores = 4)
```

```{r mdl}
# calculate mdl
mdl <- MDL(rCAM)

# format mdl table
mdl <- tibble(Data = mdl@datalengths,
        Model = mdl@mdls,
        k = mdl@K) %>%
    gather(key, val, -k)
```

```{r mdl_plot}
# plot mdl
mdl %>%
    ggplot(aes(x = k, y = val, group = key, color = key)) +
    geom_point() +
    geom_line(size = 1.2) +
    scale_y_continuous(labels = scales::scientific) + 
    labs(x = 'Number of Sources', y = 'Code Length', color = '') +
    theme_linedraw() + 
    theme(legend.position = 'top') -> mdl_plot
```

```{r extract_estimates}
# extract estimate matrices and markers
n <- 3
aset <- Amat(rCAM, n)
sset <- Smat(rCAM, n)
markers <- MGsforA(rCAM, n)
names(markers) <- paste0('P', 1:3)
```

```{r fraction_time_plot}
# plot fractions over time
aset_df <- aset %>%
    as_tibble() %>%
    setNames(paste0('P', 1:3)) %>%
    mutate(time = time) %>%
    gather(pop, frac, -time)

aset_df %>%
    ggplot(aes(x = time, y = frac, color = pop, group = pop)) +
    geom_point() +
    geom_smooth(se = FALSE) +
    labs(x = 'Time (hour)', y = 'Fraction of Cells') + 
    theme_bw() + 
    theme(legend.position = 'none',
          panel.grid = element_blank(),
          panel.border = element_rect(size = 1.5)) +
    annotate('text', x = rep(260, 3), y = c(.03, .32, .65),
             label = c('P3', 'P1', 'P2'),
             color = c('blue', 'red', 'darkgreen')) -> fraction_time
```

```{r get_simplex}
# extract simplex data
devtools::install('debCAM/')
simplex <- debCAM::simplexplot(mat,
                               aset,
                               markers,
                               plot = FALSE)

# format simplex table
simplex <- as_tibble(t(simplex)) %>%
    mutate(gene = rownames(t(simplex))) %>%
    left_join(melt(markers) %>% setNames(c('gene', 'pop'))) %>%
    mutate(pop = ifelse(!is.na(pop), paste0('P', pop), NA))
```

```{r simplex_plot}
# plot simplex
simplex %>%
    ggplot(aes(x = V1, y = V2)) +
    geom_point(color = 'gray') +
    geom_point(data = na.omit(simplex), aes(color = pop)) + 
    labs(x = 'Dimension One', y = 'Dimension Two',
         color = 'Sup-Population') +
    theme_bw() + 
    theme(legend.position = 'none',
          panel.grid = element_blank(),
          panel.border = element_rect(size = 1.5)) +
    annotate('text', x = c(-.05, .15, 0), y = c(-.25, 0, .2),
             label = c('P3', 'P1', 'P2'),
             color = c('blue', 'red', 'darkgreen')) -> simplexplot
```

```{r markers_plot}
# plot the adipogenic markers
adipo_markers <- list('Adipogenic' = c('Pparg', 'Cebpb', 'Cebpa', 'Adipoq'),
                      'Lipogenic' = c('Lpl', 'Acly', 'Fasn', 'Plin1', 'Lipe'))

# subset estimated expression by adipogenic markers
adipo_mat <- sset[rownames(sset) %in% unlist(adipo_markers),]
adipo_mat <- adipo_mat[unlist(adipo_markers),]

# rename columns
colnames(adipo_mat) <- paste0('P', 1:3)

# make heatmap
png(filename = 'manuscript/adipogenic_markers.png',
    height = 9, width = 6, units = 'cm', res = 300)

exprs_color <- colorRamp2(c(0, 20), c('white', 'darkblue'))

Heatmap(log2(adipo_mat + 1),
        col = exprs_color,
        row_split = rep(names(adipo_markers), lengths(adipo_markers)),
        show_heatmap_legend = FALSE,
        column_names_rot = 0,
        column_names_centered = TRUE)

dev.off()
```

```{r other_markers}
#other_markers <- list(
#    'Insulin' = c('Insr', 'Irs1', 'Slc2a4', 'Pdgfb'),
#    'WT1' = c('Wt1', 'Ucp1'),
#    'Keratins' = c('Krt5', 'Krt14', 'Krt15')
#                      )

other_markers <- c('Insr', 'Irs1', 'Slc2a4', 'Pdgfb')

# subset the estimated expression matrix by other marekrs
other_mat <- sset[rownames(sset) %in% other_markers,]
other_mat <- other_mat[other_markers,]

# rename the columns
colnames(other_mat) <- paste0('P', 1:3)

# make a plot
png(filename = 'manuscript/other_markers.png',
    height = 6, width = 6, units = 'cm', res = 300)

exprs_color <- colorRamp2(c(0, 15), c('white', 'darkblue'))

Heatmap(log2(other_mat + 1),
        col = exprs_color,
        show_heatmap_legend = FALSE,
        column_names_rot = 0,
        column_names_centered = TRUE)

dev.off()
```

```{r markers_stats}
# calculate markers stats (fold-change)
fold_change <- MGstatistic(mat + .01,
                             aset,
                             boot.alpha = 0.05,
                             nboot = 1000,
                             cores = 4)

# format markers statistics
markers_stats <- fold_change %>%
    rownames_to_column('gene') %>%
    dplyr::filter(gene %in% unlist(markers), OVE.FC != Inf, !grepl('Rik', gene)) %>%
    mutate(idx = paste0('P', idx),
           gene = fct_reorder(gene, OVE.FC)) %>%
    group_by(idx)
```

```{r markers_fc_plot}
markers_stats %>%
    dplyr::slice(1:5) %>%
    ggplot(aes(x = gene, y = log2(OVE.FC))) +
    geom_col() +
    scale_y_continuous(limits = c(0,6.2), expand = c(0, .1)) +
    facet_wrap(~idx, scales = 'free_x') +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1),
          strip.background = element_blank(),
          panel.spacing = unit(-.05, 'cm'),
          panel.grid = element_blank(),
          panel.border = element_rect(size = 1.5)) +
    labs(x = '', y = 'Fold-change (log_2)') -> markers_fc
```

```{r more_markersstats}
# summarizing fold-change of the markers
summary(markers_stats$OVE.FC)

markers_stats %>%
    ggplot(aes(x = log2(OVE.FC))) +
    geom_density() +
    facet_wrap(~idx)

summary(markers_stats$OVE.FC.alpha)

markers_stats %>%
    ggplot(aes(x = (OVE.FC.alpha))) +
    geom_density() +
    facet_wrap(~idx)

```

```{r more_markers}
mm <- adipo_markers
mm$other <- other_markers

melt(mm) %>%
    setNames(c('gene', 'category')) %>%
    left_join(rownames_to_column(fold_change, 'gene'))
```

```{r estimate_human}
# subset to common gene symbols
new_markers <- map(markers, function(x) {
                       intersect(toupper(x), rownames(primary_mat))
                   })

# estimate the populations fractions
set.seed(111)
rre <- redoASest(primary_mat,
                 new_markers,
                 maxIter = 10)

# format re-restimated fractions
rre_df <- as_tibble(rre$Aest) %>%
    setNames(paste0('P', 1:3)) %>%
    mutate(group = factor(group, levels = c('NDf', 'Df'))) %>%
    gather(pop, frac, -group)

rre_ave <- rre_df %>% 
    group_by(pop, group) %>%
    dplyr::summarise(frac = mean(frac))
```

```{r primary_plot}
# make a plot for the fractions in primary adipocytes
rre_df %>% 
    ggplot() +
    geom_jitter(aes(x = group, y = frac, color = pop, group = pop),
                width = .1) +
    geom_point(data = rre_ave, 
               aes(x = group, y = frac, color = pop, group = pop)) +
    geom_line(data = rre_ave, aes(x = group, y = frac, color = pop, group = pop), size = 1) +
    theme_linedraw() +
    theme(legend.position = 'top') +
    labs(x = '', y = 'Fraction of Cells') +
    theme_bw() + 
    theme(legend.position = 'none',
          panel.grid = element_blank(),
          panel.border = element_rect(size = 1.5)) +
    annotate('text', x = rep(2.3, 3), y = c(.05, .35, .62),
             label = c('P3', 'P1', 'P2'),
             color = c('blue', 'red', 'darkgreen')) +
    scale_x_discrete(labels = c('HPP', 'HPP + MDI')) -> primary_adipocytes
```

```{r distances}
# calculate distance between sub-populations
## common symbols
ind <- intersect(toupper(rownames(sset)), rownames(rre$Sest))

## rename rows
mm <- sset
rownames(mm) <- toupper(rownames(mm))

## subset matrix
mm <- mm[ind,]
hs <- rre$Sest[ind,]

## get distances
dist_models <- as.matrix(pdist(t(log2(mm + 1)), t(log2(hs + 1))))

## rename columns and rows
rownames(dist_models) <- paste0('P', 1:3)
colnames(dist_models) <- paste0('P', 1:3)

## get correlations
m1 <- as.matrix(scale(mm))
colnames(m1) <- paste('MMP', 1:3)

m2 <- as.matrix(scale(hs))
colnames(m2) <- paste('HSP', 1:3)

cor_mat <- cor(m1, m2, method = 'spearman')
colnames(cor_mat) <- paste0('P', 1:3)
rownames(cor_mat) <- paste0('P', 1:3)
```

```{r distances_plot,eval=FALSE}
# make a distance heatmap
png(filename = 'manuscript/distances.png',
    height = 5, width = 5, units = 'cm', res = 300)

dmat <- (dist_models/min(dist_models)) - 1 

dist_color <- colorRamp2(c(0, max(dmat)), c('white', 'darkblue'))

Heatmap(dmat, 
        col = dist_color,
        show_heatmap_legend = FALSE,
        column_names_rot = 0,
        column_names_centered = TRUE)

dev.off()
```

```{r correlations_plot}
# make a distance heatmap
png(filename = 'manuscript/distances.png',
    height = 5, width = 5, units = 'cm', res = 300)

cor_color <- colorRamp2(c(min(cor_mat), max(cor_mat)),
                        c('white', 'darkblue'))

Heatmap(cor_mat, 
        col = cor_color,
        show_heatmap_legend = FALSE,
        column_names_rot = 0,
        column_names_centered = TRUE)

dev.off()
```

```{r more_population_correlations}
rcorr(m1, m2, type = 'spearman') %>%
    map_df(melt, .id = 'type')
```

```{r iterations}
# run iterated estimations of fractions
iterated_rre <- rerun(100, {
    # sample markers to lowest group number 
    m <- map(markers, ~sample(.x, 35))
    
    # estimate fractions
    a <- redoASest(mat,
              m,
              maxIter = 2,
              methy = FALSE)
    
    # formate the coutpus
    a$Aest %>%
        as_tibble() %>%
        setNames(paste0('P', 1:3)) %>%
        mutate(time = adipo_counts$time)
}) %>%
    bind_rows(.id = 'iteration')
```

```{r iterations_plot}
# make iterations plot
iterated_rre %>%
    gather(pop, frac, -time,-iteration) %>%
    ggplot(aes(y = frac, x = time, group = iteration)) +
    geom_smooth() +
    facet_wrap(~pop, nrow = 1) +
    labs(x = 'Time (hour)', y = 'Fraction of Cells',
         color = 'Sub-population') + 
    theme_bw() + 
    theme(panel.spacing = unit(-.05, 'cm'),
          strip.background = element_blank(),
          panel.grid = element_blank(),
          panel.border = element_rect(size = 1.5)) -> fraction_time_iterations
```

```{r iterations_bootstrapping}
# gather estimated fractions by populations
iteractions_groups <- iterated_rre %>%
    gather(pop, frac, starts_with('P'))

# show the density of estimated fractions at different times
iteractions_groups %>%
    ggplot(aes(x = frac, group = as.factor(time))) +
    geom_density() +
    facet_wrap(~pop)

# show the histogram of the cv of estimated fractions
iteractions_stats <-  iteractions_groups %>%
    group_by(time, pop) %>%
    summarise(ave = mean(frac),
              se = var(frac),
              cv = se/ave) 

iteractions_stats %>%
    ggplot(aes(x = cv)) +
    geom_histogram() +
    facet_wrap(~pop)

# what is most cv like?
iteractions_stats %>%
    with(cv < .1) %>%
    table()

# calculate bias in bootstrapped estimates
iterations_bias <- aset_df %>%
    group_by(time, pop) %>%
    summarise(frac = mean(frac)) %>%
    ungroup() %>%
    full_join(iteractions_stats) %>%
    mutate(bias = frac - ave)

# show the distribution of bias
iterations_bias %>%
    ggplot(aes(x = abs(bias))) +
    geom_histogram() +
    facet_wrap(~pop)

iterations_bias %>%
    with(bias < .1) %>%
    table()
```

```{r deconv_pharma}
# run deconv
pharma_rre <- map(pharma_mats, function(x) {
    redoASest(x, markers)
})

# extract fractions
pharma_fractions <- map_df(pharma_rre, function(x) {
    as_tibble(x$Aest) %>%
        setNames(paste0('P', 1:3)) %>%
        mutate(sample_id = rownames(x$Aest))
}, .id = 'series_id') %>%
    gather(pop, frac, starts_with('P'))
```

## Over-representation analysis

```{r go_annotations}
# extract annotations
## get mouse gene symbols
symbols <- keys(org.Mm.eg.db, 'SYMBOL')

## get go ids to symbols
term2gene <- AnnotationDbi::select(org.Mm.eg.db,
                    symbols,
                    'GO',
                    'SYMBOL') %>%
    dplyr::select(term = GO, gene = SYMBOL) %>%
    unique()

## get go terms
terms <- as.data.frame(GOTERM)
terms <- tibble(ID = terms$go_id,
                term = terms$Term,
                definition = terms$Definition,
                ontology = terms$Ontology)

## intersect markers symbols with go symbols
markers_symbols <- map(markers, intersect, y = unique(term2gene$gene))
str(markers_symbols)
```

```{r over_representation}
# run over-representation test
comp <- map_df(markers_symbols,
            function(x) {
                enricher(x,
                         TERM2GENE = term2gene,
                         pAdjustMethod = 'fdr')@result
            }, .id = 'population')
```

```{r}
# set.seed(123)
# comp2 <- compareCluster(markers_symbols,
#                         fun = 'enricher',
#                         TERM2GENE = term2gene,
#                          pAdjustMethod = 'fdr')
# comp2@compareClusterResult %>%
#     inner_join(terms) %>%
#     as_tibble() %>%
#     dplyr::select(Cluster, ontology, term, pvalue, p.adjust, qvalue, Count) %>%
#     dplyr::filter(p.adjust < .2) %>%
#     unique() %>%
#     arrange(Cluster, ontology, desc(Count)) %>%
#     write_csv('manuscript/tables/marker_enrichment.csv')
```

```{r more_ova}
# inner_join(terms, comp) %>%
#     dplyr::filter(pvalue < .05, Count > 3) %>%
#     dplyr::select(population, ontology, term, pvalue, p.adjust, qvalue) %>%
#     unique() %>%
#     arrange(population, ontology) %>%
#     write_csv('manuscript/tables/marker_enrichment.csv')
```

```{r gsea}
ove_fc <- MGstatistic(sset + 1,
                      A = paste0('P', 1:3),
                      boot.alpha = 0.05,
                      cores = 4)

ove_fc2 <- map2(split(ove_fc$OVE.FC, ove_fc$idx), 
               split(rownames(ove_fc), ove_fc$idx),
               function(vec, nm) {
                   names(vec) <- nm
                   vec
               })

str(ove_fc2)

go_ids <- dplyr::filter(comp, p.adjust < .2, Count > 2) %>%
    pull(ID) %>%
    unique()

go_terms <- AnnotationDbi::select(org.Mm.eg.db,
                                  go_ids,
                                  'SYMBOL',
                                  'GO') %>%
    with(split(SYMBOL, GO)) %>%
    map(unique)

comp_pops <- map_df(ove_fc2, function(x) {
    ind <- c('GO:0050873', 'GO:0004879', 'GO:0003707')
    fgsea(go_terms[ind],
          x,
          nperm = 1000,
          nproc = 4)
}, .id = 'pop') %>%
    left_join(terms, by = c('pathway'='ID'))
```

## Differential expression analysis

```{r occupancy}
# # get counts in peaks
# peak_counts <- read_rds('data/peak_counts.rds')
# 
# # subset to POLR2A
# se_pol <- peak_counts[, grepl('POLR2A', peak_counts$factor)]
# 
# # aggregate by gene
# mat_pol <- assay(se_pol) %>%
#     as_tibble() %>%
#     mutate(gene = as.character(mcols(se_pol)$geneId)) %>%
#     group_by(gene) %>%
#     summarise_all(sum) %>%
#     ungroup() %>%
#     column_to_rownames('gene')
```

```{r estimate_fractions_occupance}
# # estimate fractions based on occupancy
# rre_pol <- redoASest(mat_pol, markers, maxIter = 10)
```

```{r population_markers}
# cell cycle index genes
mat_pop <-  assay(adipo_counts)[c('Ank2', 'Aldoc', 'Gapdh'),
                               adipo_counts$time %in% c(0:10 * 24)]
tt <- adipo_counts$time[adipo_counts$time %in% c(0:10 * 24)]

as_tibble(t(mat_pop)) %>%
    mutate(time = tt) %>%
    group_by(time) %>%
    summarise_all(mean)
```

```{r occupancy_fractions}
# rre_pol$Aest %>%
#     as_tibble() %>%
#     setNames(paste0('P', 1:3)) %>%
#     mutate(time = se_pol$time) %>%
#     gather(pop, frac, -time) %>%
#     ggplot(aes(x = time, y = frac, color = pop, group = pop)) +
#     geom_point() +
#     geom_smooth(se = FALSE) +
#     labs(x = 'Time Point (hour)', y = 'Fraction of Cells',
#          color = 'Sub-population') + 
#     theme_linedraw() + 
#     theme(legend.position = 'top') -> fraction_time_trans
```


```{r cc_index}
# define calculate index function
calculate_index <- function(mat, pos, neg = NULL, n) {
    vecs <- list(pos = pos, neg = neg) %>%
        map(function(x) {
            ind <- rownames(mat) %in% x
            m <- mat[ind,]
            m
        })
    if(!is.null(neg)) {
        (colSums(vecs$pos) - vecs$neg)/n
    } else {
        colSums(vecs$pos)/n
    }
}

# cell cycle index genes
indecies <- list(
    pos = c('Mki67', 'Rb1', 'Hist1h2ae', 'Ccnb1', 'Cbx3', 'Gapdh', 'Ccnb2'),
    neg = c('E2f1')
)
mat <- assay(dds_transform)
time <- dds_transform$time

cc_index <- tibble(
    t = time[time %in% c(0, 24, 48)],
    i = calculate_index(
        mat[,time %in% c(0, 24, 48)],
        indecies$pos,
        indecies$neg,
        8)
)
```

```{r cc_index_ecdf}
cc_index %>%
    ggplot(aes(x = scales::rescale(i, to = c(0, 1)), color = as.factor(t))) +
    stat_ecdf(size = 1.5) +
    theme_bw() +
    theme(legend.position = c(.2, .8),
          panel.grid = element_blank(),
          legend.background = element_blank(),
          panel.border = element_rect(size = 1.5)) +
    labs(x = 'Cell Cycle Index',
         y = 'Cumulative Distribution',
         color = '') -> cc_index_ecdf
```

```{r more_ecdf}
cc_groups <- split(cc_index$i, cc_index$t)

ks.test(cc_groups$`24`, cc_groups$`0`, alternative = 'less')
ks.test(cc_groups$`24`, cc_groups$`48`, alternative = 'less')
```

```{r cc_expression}
cc_expression <- mat[rownames(mat) %in% unlist(indecies), time %in% c(0, 24, 48)] %>%
    melt() %>%
    as_tibble() %>%
    setNames(c('gene', 'id', 'count')) %>%
    inner_join(tibble(id = dds$id, time = as.factor(dds$time)))
```

```{r more_cc_expression}
# run deseq with time as a factor
dds_time <- DESeqDataSet(se, ~as.factor(time))
dds_time <- DESeq(dds_time)

# get fold change for all genes
stats <- list('0' = 'as.factor.time.0',
              '48' = 'as.factor.time.48') %>%
    map(function(x) {
        res <- results(dds_time,
                        contrast = list('as.factor.time.24', x),
                        tidy = TRUE)
        vec <- res$log2FoldChange
        names(vec) <- res$row
        vec
    })

# calculate enrichment of cell cycle indicies
set.seed(12345)
map_df(stats,
    function(x) {
        fgsea::fgsea(list('Cell Cycle' = unlist(indecies)),
                     x,
                     nperm = 1000,
                     nproc = 4)
    }, .id = 'time')
```

```{r cc_index_expression}
cc_expression %>%
    ggplot(aes(x = time, y = rescale(count, c(0, 1)))) +
    geom_boxplot(notch = TRUE, size = 1) +
    labs(x = 'Time (hr)',
         y = 'Expression of Cell Cycle Genes') +
    theme_bw() +
    theme(panel.grid = element_blank(),
          panel.border = element_rect(size = 1.5)) -> cc_index_expression
```

```{r cc_population_specific}
# population specific expression of cell cycle index genes
models <- map(list('24 vs 0 (hr)' = c(0, 24),
                   '48 vs 24 (hr)' = c(24, 48)), function(t) {
    ind <- dds$time %in% t
    se <- dds[, ind]
    se <- se[rowMeans(assay(se)) > 0,]
    fac <- ifelse(se$time == t[1], 0, 1)
    
    refs <- map(markers, function(x) marker(assay(se), intersect(rownames(se), x)))
    diffs <- map(refs, function(x) fac * x)
    
    gens <- unlist(indecies)
    names(gens) <- unlist(indecies)

    list(
        P1 = map(gens,function(i) lm(assay(se)[i,] ~ refs$P1 + diffs$P1)),
        P2 = map(gens,function(i) lm(assay(se)[i,] ~ refs$P2 + diffs$P2))
    )
})

cr <- map_df(models,
             function(t) map_df(t,
                                function(p) map_df(p,
                                                   function(p) {
    tibble(cr = as.numeric(residuals(p) + as.matrix(p$model[,-1]) %*% p$coef[-1]),
           ref = p$model[, 2],
           diff = p$model[, 3])
}, .id = 'gene'),.id = 'pop'), .id = 'time')
```

```{r cc_psea}
cr %>%
    ggplot(aes(x = ref, y = rescale(cr))) +
    geom_point(size = .5, color = 'darkgray') +
    geom_smooth(method = 'lm', se = FALSE) +
    facet_grid(pop ~ time) +
    labs(x = 'Population Specific Expression',
         y = 'Partial Residuals') +
    scale_y_continuous(breaks = seq(0, 1, .2), labels = seq(0, 1, .2)) +
    theme_bw() +
    theme(panel.grid = element_blank(),
          panel.border = element_rect(size = 1.5),
          strip.background = element_blank(),
          strip.text.y = element_text(angle = 0),
          panel.spacing = unit(-.1, 'cm')) -> cc_psea
```

```{r differential_expresison}
# mixture differential expression
pharma_de <- map2_df(pharma_mats, selected_list, function(x, y) {
    # make factor of the treatment type
    fac <- factor(y$treatment_type)
    
    # make a model matrix
    mod <- model.matrix(~ fac)
    colnames(mod) <- levels(fac)
    
    # fit by the model
    fit <- lmFit(log2(x + 1), mod)
    fit <- eBayes(fit)
    
    coeff <- colnames(mod)[-1]
    names(coeff) <- coeff
    
    # extract toptable for each coefficient
    map_df(coeff, function(f) {
        topTable(fit,
                 coef = f,
                 number = Inf,
                 genelist = rownames(x)) %>%
            as_tibble() %>%
            dplyr::rename(gene_id = ID)
    }, .id = 'treatment_type')
}, .id = 'series_id')
```

```{r populations_specific}
# population specific expression analysis
pharma_psa <- map2_df(pharma_mats, selected_list, function(m, f) {
    # extract treatment types other than control
    trt <- f$treatment_type[!grepl('control', f$treatment_type)]
    
    # map over treatment types
    map_df(trt, function(x) {
        # subset expression matrix
        ind <- f$treatment_type %in% c('control', x)
        mm <- m[, ind]
        
        # make a treatment vs control factor
        fac <- as.numeric(factor(f$treatment_type[ind])) - 1
        
        # map over the sub-populations
        map_df(markers, function(r, .y) {
            # extract the reference expression
            ref <- PSEA::marker(mm, intersect(rownames(mm), r))
            
            # factor the treament vs control in the reference
            dif <- fac * ref
            
            # make a model matrix and fit
            mod <- model.matrix(~ ref + dif)
            fit <- limma::lmFit(log2(mm + 1), mod)
            
            # make a contrast matrix and fit
            cont.mat <- makeContrasts(dif-ref, levels = mod)
            fit2 <- contrasts.fit(fit, cont.mat)
            fit2 <- limma::eBayes(fit2)
            
            # extract top tables
            topTable(fit2, coef = 1, number = Inf, genelist = rownames(mm)) %>%
                mutate(treatment_type = x)
        }, .id = 'pop')
    })
})
```


```{r correlating_mixs_pops}
## similariy of sub-populations to the mixtures
mix_pops <- map(c('rosiglitazone', 'ibuprofen',  'stearic acid'), function(x) {
    df1 <- filter(pharma_de,
                  treatment_type == x) %>%
        slice_max(abs(logFC), n = 500)
    
    df2 <- filter(pharma_psa,
                  treatment_type == x & ID %in% df1$gene_id)
    
    bind_rows(df1, df2) %>%
        mutate(pop = ifelse(is.na(pop), 'Mix', pop))
})

names(mix_pops) <- c('rosiglitazone', 'ibuprofen',  'stearic acid')

## correlation between fold-change in response to drugs
treatment_fc <- acast(pharma_de, gene_id ~ treatment_type, value.var = 'logFC')
treatment_cor <- cor(scale(treatment_fc), use = 'complete')
```

```{r more_perturbations}
# test correlations among familes of compounds
rcorr(treatment_fc) %>%
    map_df(melt, .id = 'type')

# test differences in ecdfs
mix_pop_groups <- map(mix_pops, function(x) split(x$logFC, x$pop))

# rosi
ks.test(mix_pop_groups$rosiglitazone$P2,
        mix_pop_groups$rosiglitazone$Mix,
        alternative = 'less')

ks.test(mix_pop_groups$rosiglitazone$P2,
        mix_pop_groups$rosiglitazone$P1,
        alternative = 'less')

ks.test(mix_pop_groups$rosiglitazone$P2,
        mix_pop_groups$rosiglitazone$P3,
        alternative = 'less')

# ibu
ks.test(mix_pop_groups$ibuprofen$P2,
        mix_pop_groups$ibuprofen$Mix,
        alternative = 'less')

ks.test(mix_pop_groups$ibuprofen$P2,
        mix_pop_groups$ibuprofen$P1,
        alternative = 'less')

ks.test(mix_pop_groups$ibuprofen$P2,
        mix_pop_groups$ibuprofen$P3,
        alternative = 'less')

# stearic
ks.test(mix_pop_groups$ibuprofen$P1,
        mix_pop_groups$ibuprofen$Mix,
        alternative = 'less')

ks.test(mix_pop_groups$ibuprofen$P1,
        mix_pop_groups$ibuprofen$P2,
        alternative = 'less')

ks.test(mix_pop_groups$ibuprofen$P1,
        mix_pop_groups$ibuprofen$P3,
        alternative = 'less')
```

```{r format_perturbation_data}
adipo_markers_df <- tibble(
    marker_type = rep(c('Adipogenic', 'Lipogenic'), each = 4),
    gene_id = c('Pparg', 'Cebpb', 'Cebpa', 'Adipoq', 'Lpl', 'Acly', 'Fasn', 'Lipe')
) %>%
    arrange(gene_id)

hmb <- pharma_de %>%
    inner_join(adipo_markers_df) %>%
    inner_join(dplyr::select(selected_samples, -sample_id)) %>%
    dcast(treatment_category + treatment_type ~ gene_id, value.var = 'logFC', fun.aggregate = mean) %>%
    with({
        m <- dplyr::select(., -starts_with('treatment')) %>% as.matrix()
        rownames(m) <- .$treatment_type
        Heatmap(scale(m),
                name = 'Fold-change',
                #col = de_color,
                cluster_rows = FALSE,
                cluster_row_slices = FALSE,
                column_title = NULL,
                cluster_columns = FALSE,
                cluster_column_slices = FALSE,
                row_split = .$treatment_category,
                column_split = adipo_markers_df$marker_type,
                heatmap_legend_param = list(direction = "horizontal"),
                show_heatmap_legend = FALSE)
    })

psa <- as_tibble(pharma_psa) %>%
    dplyr::select(gene_id = ID, everything()) %>%
    inner_join(adipo_markers_df) %>%
    inner_join(dplyr::select(selected_samples, -sample_id)) %>%
    dcast(treatment_category + treatment_type + pop ~ gene_id,
          value.var = 'logFC',
          fun.aggregate = mean) %>%
    with({
        m <- dplyr::select(., 4:ncol(.)) %>% as.matrix()
        rownames(m) <- ifelse(duplicated(.$treatment_type),
                              .$pop, 
                              paste(.$pop, .$treatment_type, sep = '_'))
        Heatmap(scale(m),
                name = 'Fold-change (log_2)',
                row_split = .$treatment_category,
                cluster_rows = FALSE,
                cluster_row_slices = FALSE,
                column_title = NULL,
                cluster_columns = FALSE,
                cluster_column_slices = FALSE,
                column_split = adipo_markers_df$marker_type,
                heatmap_legend_param = list(direction = "horizontal"),
                show_heatmap_legend = FALSE)
    })
```

## Population dynamics analysis

```{r toy_model}
toy_model <- ode(times = seq(1:48),                 # times
    parms = c(r = .05),                # parameters
    y = c(P1 = 90,
          P2 = 10,
          N = 100),# states
    func = function(t, state, parms) {
        with(as.list(c(parms, state)), {
            dP1 = -r * P1
            dP2 = r * P2
            dN = (P2-P1) * r
            return(list(c(dP1, dP2, dN)))
        })
    }) %>%
    as.data.frame() %>%
    gather(pop, n, -time)
```

```{r explore_models}
# models with diffent starting parameters
map(c(.01, .02, .03, .04), function(x) {
    ode(times = seq(1:48),                 # times
    parms = c(r = x),               # parameters
    y = c(P1 = 90,
          P2 = 10,
          N = 100),# states
    func = function(t, state, parms) {
        with(as.list(c(parms, state)), {
            dP1 = -r * P1
            dP2 = r * P2
            dN = (P2-P1) * r
            return(list(c(dP1, dP2, dN)))
        })
    }) %>%
    as.data.frame() %>%
    gather(pop, n, -time) %>%
    ggplot(aes(x = time, y = n, color = pop)) +
    geom_line(size = 1) +
    theme_bw() +
    labs(x = 'Time (hr)',
         y = 'Number of Cells',
         color = '') +
    ggtitle(paste('r', x, sep = ' = '))
}) %>%
    plot_grid(plotlist = .,
              ncol = 2)
```


```{r explore_models}
# models with diffent starting states
map(c(.1, .2, .3, .4), function(x) {
    N <- 100
    P1 <- 100 * (1-x)
    P2 <- 100 * x
    
    ode(times = seq(1:48),                 # times
    parms = c(r = .025),               # parameters
    y = c(P1 = P1,
          P2 = P2,
          N = N),# states
    func = function(t, state, parms) {
        with(as.list(c(parms, state)), {
            dP1 = -r * P1
            dP2 = r * P2
            dN = (P2-P1) * r
            return(list(c(dP1, dP2, dN)))
        })
    }) %>%
    as.data.frame() %>%
    gather(pop, n, -time) %>%
    ggplot(aes(x = time, y = n, color = pop)) +
    geom_line(size = 1) +
    theme_bw() +
    labs(x = 'Time (hr)',
         y = 'Number of Cells',
         color = '') +
    ggtitle(paste('P2/N', x, sep = ' = '))
}) %>%
    plot_grid(plotlist = .,
              ncol = 2)
```

```{r toy_model_plot}
 toy_model %>%
    ggplot(aes(x = time, y = n, color = pop)) +
    geom_line(size = 1) +
    theme_bw() +
    labs(x = 'Time (hr)',
         y = 'Number of Cells',
         color = '') +
    theme(legend.position = c(.5, .8),
          panel.grid = element_blank(),
          legend.background = element_blank(),
          panel.border = element_rect(size = 1.5)) -> toy_model_plot
```

## Generating figures and tables
## Tables

```{r tables_dir}
dir.create('manuscript/tables')
```

```{r enrichment_table}
# merge annotation and enrichment
inner_join(terms, comp) %>%
    dplyr::filter(p.adjust < .2, Count > 2) %>%
    unique() %>%
    group_by(population, ontology) %>%
    write_csv('manuscript/tables/marker_enrichment.csv')
```

```{r rnaseq_dataset}
# adipo_counts
colData(adipo_counts) %>%
    as_tibble() %>%
    group_by(study, bibtexkey) %>%
    summarise(time = paste0(min(time), ':', max(time))) %>%
    dplyr::select(study, time, bibtexkey) %>%
    mutate(bibtexkey = paste0('\\cite{', bibtexkey, '}')) %>%
    setNames(c('SRA ID', 'Time (hrs)', 'Ref.' )) %>%
    xtable(align = 'clcc') %>%
    print(floating = FALSE,
          include.rownames = FALSE,
          booktabs = TRUE,
          sanitize.text.function = identity,
          comment = FALSE,
          file = 'manuscript/tables/counts_data.tex')
```

```{r microarrays_perturbation_dataset}
# microarrays dataset table
read_csv('data/perturbations_md.csv') %>%
    arrange(treatment_type) %>%
    group_by(series_id, citation) %>%
    summarise(cat = unique(treatment_category)[-1],
              treatment = paste(unique(treatment_type)[-1], collapse = '/ ')) %>%
    dplyr::select(series_id, cat, treatment, citation) %>%
    mutate(citation = paste0('\\cite{', citation, '}')) %>%
    setNames(c('GEO ID', 'Type', 'Treatment', 'Ref.' )) %>%
    xtable(align = 'cllp{.2\\textwidth}c') %>%
    print(floating = FALSE,
          include.rownames = FALSE,
          booktabs = TRUE,
          sanitize.text.function = identity,
          comment = FALSE,
          file = 'manuscript/tables/pharma_data.tex')
```

## Figures

```{r figures_dir}
dir.create('manuscript/figures')
```

```{r subpopulations}
plot_grid(mds,
          simplexplot,
          fraction_time,
          primary_adipocytes,
          ncol = 2,
          scale = .9,
          labels = 'AUTO',
          label_fontface = 'plain') %>%
    ggsave(plot = .,
           filename = 'manuscript/figures/subpopulations.png',
           height = 15, width = 15, units = 'cm')
```

```{r markers_iterations}
plot_grid(markers_fc,
          fraction_time_iterations,
          ncol = 1,
          scale = .9,
          labels = 'AUTO',
          label_fontface = 'plain') %>%
    ggsave(plot = .,
           filename = 'manuscript/figures/markers_iterations.png',
           height = 15, width = 12, units = 'cm')
```

```{r heatmaps.png}
png(filename = 'manuscript/figures/heatmaps.png',
    height = 10, width = 13, units = 'cm', res = 300)

grid.arrange(arrangeGrob(rasterGrob(readPNG('manuscript/adipogenic_markers.png')),
                         top = textGrob("A", x = unit(0, "npc"), y = unit(1, "npc"), just=c("left","top"))),
             arrangeGrob(
                 arrangeGrob(rasterGrob(readPNG('manuscript/other_markers.png')),
                             top = textGrob("B", x = unit(0, "npc"), y = unit(1, "npc"), just=c("left","top"))),
                 arrangeGrob(rasterGrob(readPNG('manuscript/distances.png')),
                             top = textGrob("C", x = unit(0, "npc"), y = unit(1, "npc"), just=c("left","top"))),
             ncol = 1),
             nrow = 1)
dev.off()

# remove individual heatmaps files
unlink('manuscript/adipogenic_markers.png')
unlink('manuscript/other_markers.png')
unlink('manuscript/distances.png')
```

```{r cell_cycle.png}
plot_grid(toy_model_plot,
          cc_index_ecdf, 
          cc_index_expression, 
          cc_psea,
          ncol = 2,
          scale = .9,
          labels = 'AUTO',
          label_fontface = 'plain') %>%
ggsave(plot = .,
       filename = 'manuscript/figures/cell_cycle.png',
       height = 15, width = 16, units = 'cm')
```

```{r perturbation_plot,eval=FALSE}
# make a plot
png(filename = 'manuscript/figures/pharma_perturbations.png',
    width = 12, height = 30, units = 'cm', res = 300)

draw(psa %v% hmb,
     heatmap_legend_side = "bottom", 
     annotation_legend_side = "bottom")

dev.off()
```

```{r treatment_deg}
pps <- mix_pops %>%
    map(function(x) {
        x  %>%
            ggplot(aes(x = abs(logFC), color = pop)) +
            stat_ecdf() +
            theme_bw() +
            theme(legend.position = c(.7, .5),
                  panel.grid = element_blank(),
                  panel.border = element_rect(size = 1.5)) +
            labs(x = 'Fold-change',
                 y = 'Cumulative Distribution',
                 color = '')
    })

# correlation between mixtures
order_ind <- unique(pharma_de$treatment_type)
ggcorr(treatment_cor[order_ind, order_ind],
       hjust = 1,
       palette = 'RdYlBu',
       low = 'blue',
       high = 'red',
       legend.position = 'none',
       label_size = .5,
       label_color = 'gray') +
    theme(plot.margin=grid::unit(c(0,0,0,0), "mm")) -> pps[[4]]

plot_grid(pps[[4]], pps[[1]], pps[[2]], pps[[3]],
          nrow = 2,
          scale = .9,
          labels = 'AUTO',
          label_fontface = 'plain') %>%
    ggsave(plot = .,
           filename = 'manuscript/figures/treatment_deg.png',
           width = 15, height = 15, units = 'cm')
```

```{r}
## locate and read raw ct data
dat <- read.delim('data/populations_ct.tsv')

# add grouping variable
group_var <- factor(dat$group * 24)
ct <- dat[, -1]

# calculate all values and errors in one step
gg <- lapply(c('Ank2', 'Aldoc', 'Pparg', 'Cebpb'), function(x) {
  df <- data.frame(ct[, x], ct[, 'Rn18s'])
  names(df) <- c(x, 'Rn18s')
  res <- pcr_analyze(df,
                     group_var = group_var,
                     reference_gene = 'Rn18s',
                     reference_group = '-48',
                     method = 'delta_delta_ct')
  
  ggplot(res, aes(x = group , y = relative_expression)) +
    geom_col() +
    labs(x = 'Time (hr)',
         y = paste('Relative', x, 'mRNA')) +
    theme_bw() +
    theme(legend.position = c(.7, .5),
          panel.grid = element_blank(),
          panel.border = element_rect(size = 1.5))
})

plot_grid(plotlist = gg,
          nrow = 2,
          scale = .9,
          labels = 'AUTO',
          label_fontface = 'plain') %>%
    ggsave(plot = .,
           filename = 'manuscript/figures/markers_mrna.png',
           height = 12, width = 12, units = 'cm', dpi = 300)
```

```{r save}
# save
#save.image(file = 'analysis_outputs.rda')
```

```{r new_subpopulations}
mds <- mds_df %>%
    ggplot(aes(x = scale(V1), y = scale(V2), color = stage)) +
    geom_point() +
    #geom_encircle() +
    labs(x = 'Dimension One',
         y = 'Dimension Two',
         color = 'Differentiation') +
    theme(legend.position = 'none',
          panel.grid = element_blank()) +
    annotate('text', -1.1, 1.1, label = 'S0', color = '#F8766D') +
    annotate('text', 1, 1, label = 'S1', color = '#619CFF') +
    annotate('text', -.5, -2, label = 'S2', color = '#00BA38')

# plot simplex
simplexplot <- simplex %>%
    ggplot(aes(x = V1, y = V2)) +
    geom_point(color = 'gray', size = .5) +
    geom_point(data = na.omit(simplex), aes(color = pop), size = .5) + 
    labs(x = 'Dimension One', y = 'Dimension Two',
         color = 'Sup-Population') +
    theme(legend.position = 'none',
          panel.grid = element_blank()) +
    annotate('text',
             x = c(-.2, .1, .03),
             y = c(-.2, -.1, .18),
             label = c('P3', 'P1', 'P2'),
             color = c('#619CFF', '#F8766D', '#00BA38'))

fraction_time <- aset_df %>%
    ggplot(aes(x = time, y = frac, color = pop, group = pop)) +
    geom_point(size = .7) +
    geom_smooth(se = FALSE) +
    labs(x = 'Time (hour)', y = 'Fraction of Cells') + 
    theme(legend.position = 'none',
          panel.grid = element_blank()) +
    scale_x_continuous(breaks = c(-96, 0, 48, 144, 240),
                       limits = c(-96, 290)) +
    scale_y_continuous(breaks = c(0, .2, .4, .6, .8, 1)) +
    annotate('text', x = rep(275, 3), y = c(.03, .32, .65),
             label = c('P3', 'P1', 'P2'),
             color = c('#619CFF', '#F8766D', '#00BA38'))

# make a plot for the fractions in primary adipocytes
rre_df %>% 
    ggplot() +
    geom_jitter(aes(x = group, y = frac, color = pop, group = pop),
                width = .2,
                size = .7) +
    geom_point(data = rre_ave, 
               aes(x = group, y = frac, color = pop, group = pop)) +
    geom_line(data = rre_ave, aes(x = group, y = frac, color = pop, group = pop), size = 1) +
    theme(legend.position = 'top') +
    labs(x = '', y = 'Fraction of Cells') +
    theme(legend.position = 'none',
          panel.grid = element_blank()) +
    annotate('text', x = rep(2.4, 3), y = c(.05, .35, .62),
             label = c('P3', 'P1', 'P2'),
             color = c('#619CFF', '#F8766D', '#00BA38')) +
    scale_y_continuous(breaks = c(0, .2, .4, .6, .8, 1)) +
    scale_x_discrete(labels = c('HPP', 'HPP + MDI')) -> primary_adipocytes

plot_grid(mds,
          simplexplot,
          fraction_time,
          primary_adipocytes,
          ncol = 2,
          scale = .95,
          labels = 'AUTO',
          label_fontface = 'plain') %>%
    ggsave(plot = .,
           filename = 'manuscript/figures/subpopulations.png',
           height = 4.5, width = 4.5, units = 'in')
```

```{r new_markers_iterations}
markers_fc <- markers_stats %>%
    dplyr::slice(1:5) %>%
    ggplot(aes(x = gene, y = log2(OVE.FC))) +
    geom_col() +
    scale_y_continuous(limits = c(0,6.2), expand = c(0, .1)) +
    facet_wrap(~idx, scales = 'free_x') +
    theme(axis.text.x = element_text(angle = 45, hjust = 1),
          strip.background = element_blank(),
          panel.spacing = unit(1, 'mm'),
          axis.title.x = element_text(size = 0),
          panel.grid = element_blank()) +
    labs(x = '', y = 'Fold-change (log_2)')

fraction_time_iterations <- iterated_rre %>%
    gather(pop, frac, -time,-iteration) %>%
    ggplot(aes(y = frac, x = time, group = iteration)) +
    geom_smooth() +
    facet_wrap(~pop, nrow = 1) +
    labs(x = 'Time (hour)', y = 'Fraction of Cells',
         color = 'Sub-population') + 
    theme(panel.spacing = unit(1, 'mm'),
          strip.background = element_blank(),
          panel.grid = element_blank())

plot_grid(markers_fc,
          fraction_time_iterations,
          ncol = 1,
          scale = .95,
          labels = 'AUTO',
          label_fontface = 'plain') %>%
    ggsave(plot = .,
           filename = 'manuscript/figures/markers_iterations.png',
           height = 5, width = 4, units = 'in')
```

```{r new_treatment_deg}

order_ind <- unique(pharma_de$treatment_type)
corr <- ggcorr(treatment_cor[order_ind, order_ind],
       hjust = 10,
       palette = 'RdYlBu',
       low = 'blue',
       high = 'red',
       legend.position = 'none',
       label_color = 'gray') +
    theme(axis.text.y = element_text(size = 0))

order_ind

drug_response <- bind_rows(mix_pops, .id = 'drug') %>%
    ggplot(aes(x = pop, y = (logFC)), alpha = .1) +
    geom_boxplot(notch = TRUE) +
    lims(y = c(-25, 25)) +
    facet_grid(~drug) +
    labs(x = '', y = 'Fold-change (log_2)') +
    theme(panel.grid = element_blank(),
          strip.background = element_blank(),
          axis.title.x = element_text(size = 0))

plot_grid(
    plot_grid(NULL, corr, labels = c('A', ''), label_fontface = 'plain', nrow = 1, scale = .95),
    plot_grid(drug_response, labels = 'B', label_fontface = 'plain', scale = .95),
    ncol = 1
) %>%
    ggsave(plot = .,
           filename = 'manuscript/figures/treatment_deg.png',
           width = 5, height = 4.5, units = 'in')
```

```{r new_cell_cycle}

toy_model_plot <- toy_model %>%
    ggplot(aes(x = time, y = n, color = pop)) +
    geom_line(size = 1) +
    labs(x = 'Time (hr)',
         y = 'Number of Cells',
         color = '') +
    scale_x_continuous(breaks = c(0, 12, 24, 36, 48)) +
    scale_y_continuous(breaks = c(0, 25, 50, 75, 100), limits = c(0, 115)) +
    theme(legend.position = c(.5, .8),
          legend.box.background = element_blank(),
          panel.grid = element_blank(),
          legend.background = element_blank())

cc_index_ecdf <- cc_index %>%
    ggplot(aes(x = scales::rescale(i, to = c(0, 1)), color = as.factor(t))) +
    stat_ecdf(size = 1.5) +
    theme(legend.position = c(.2, .85),
          legend.box.background = element_blank(),
          panel.grid = element_blank(),
          legend.background = element_blank()) +
    labs(x = 'Cell Cycle Index',
         y = 'Cumulative Distribution',
         color = '')

cc_index_expression <- cc_expression %>%
    ggplot(aes(x = time, y = rescale(count, c(0, 1)))) +
    geom_boxplot(notch = TRUE, size = 1) +
    labs(x = 'Time (hr)',
         y = 'Expression (Cell Cycle)') +
    theme(panel.grid = element_blank())

cc_psea <- cr %>%
    ggplot(aes(x = ref, y = rescale(cr))) +
    geom_point(size = .5, color = 'darkgray') +
    geom_smooth(method = 'lm', se = FALSE) +
    facet_grid(pop ~ time) +
    scale_y_continuous(breaks = c(0, .3, .6, .9)) +
    labs(x = 'Expression (Populations)',
         y = 'Partial Residuals') +
    theme(panel.grid = element_blank(),
          strip.background = element_blank(),
          strip.text.y = element_text(angle = 0),
          panel.spacing = unit(1, 'mm'))

plot_grid(toy_model_plot,
          cc_index_ecdf, 
          cc_index_expression, 
          cc_psea,
          ncol = 2,
          scale = .95,
          labels = 'AUTO',
          label_fontface = 'plain') %>%
    ggsave(plot = .,
           filename = 'manuscript/figures/cell_cycle.png',
           height = 5, width = 5, units = 'in')

```

